aiorun.py

```diff
@@ -318,22 +318,27 @@ def sep():
         return tasks, do_not_cancel
 
     tasks, do_not_cancel = sep()
-    # Here's a protip: if you group a bunch of tasks, and some of them
-    # get cancelled, and they DON'T HANDLE THE CANCELLATION, then the
-    # raised CancelledError will bubble up to, and stop the
-    # loop.run_until_complete() line: meaning, not all the tasks in
-    # the gathered group will actually be complete. You need to
-    # enable this with the ``return_exceptions`` flag.
-    kwargs = dict(
-        return_exceptions=True
-    )
-    if sys.version_info < (3, 10):
-        kwargs['loop'] = loop
-    group = gather(*tasks, *do_not_cancel, **kwargs)
-    logger.info("Running pending tasks till complete")
-    # TODO: obtain all the results, and log any results that are exceptions
-    # other than CancelledError. Will be useful for troubleshooting.
-    loop.run_until_complete(group)
+     
+    # Only run `gather` if there are tasks to be cancelled.
+    # Following applies to >= Py 3.10:
+    # Otherwise it's going to call `events._get_event_loop()`, which
+    # is going to raise an exception, since at this point the 
+    # loop has been stopped.
+    if tasks or do_not_cancel:
+        # Here's a protip: if you group a bunch of tasks, and some of them
+        # get cancelled, and they DON'T HANDLE THE CANCELLATION, then the
+        # raised CancelledError will bubble up to, and stop the
+        # loop.run_until_complete() line: meaning, not all the tasks in
+        # the gathered group will actually be complete. You need to
+        # enable this with the ``return_exceptions`` flag.
+        if sys.version_info < (3, 10):
+            group = gather(*tasks, *do_not_cancel, return_exceptions=True, loop=loop)
+        else:
+            group = gather(*tasks, *do_not_cancel, return_exceptions=True)
+        logger.info("Running pending tasks till complete")
+        # TODO: obtain all the results, and log any results that are exceptions
+        # other than CancelledError. Will be useful for troubleshooting.
+        loop.run_until_complete(group)
```

Comment:
Thanks again for raising the issue.  I looked at it and I think the following is a better change overall (and I should have done it this way the first time round):

```suggestion
    async def wait_for_cancelled_tasks():
        return await gather(*tasks, *do_not_cancel, return_exceptions=True)

    if tasks or do_not_cancel:
        logger.info("Running pending tasks till complete")
        # TODO: obtain all the results, and log any results that are exceptions
        # other than CancelledError. Will be useful for troubleshooting.
        loop.run_until_complete(wait_for_cancelled_tasks())
```

What are your thoughts on this?

Reply:
Sure, seems good to me.
I guess now you don't need to pass `loop` to `gather` even in Py < 3.10 ?

Reply:
Yes exactly.  Are you happy for me to make a PR myself and close this one, or do you intend to change yours (this PR) to look like above?

Reply:
Should I just commit your suggestion?